Prevent duplicate orders with a single payment intent

htdat opened the pull request:
Fixes #5786

#### Changes proposed in this Pull Request

- Add a new check to compare the current processing order vs the metadata order_id value. In case, they're not matched, the process should be stopped right away. 
- There is another approach here https://github.com/Automattic/woocommerce-payments/pull/6493, which was originally proposed in #5786 . However, I discarded that approach because of the following reasons: 
1. It requires an extra GET request before updating intentions. We're trying to solve this issue for a small percentage of total payment requests (I estimate that it's less than 0.01%) so it's not worth this kind of trade-off. 
2. There are more than one code flow to place this check and ensure it's working properly. 

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

Follow test steps in #5786, and confirm this issue does not happen.

**Note**: 

- You might check the browser dev network tab, then refresh the check out page a few times. You should start opening two tabs only until there is no AJAX / XHR request to create a new payment intent. That is, the checkout process will use the cache intent saved in the session.
- AJAX requests for creating payment intent: 
    - Split UPE http://localhost:8082/?wc-ajax=wcpay_create_payment_intent_card
    - Legacy UPE: http://localhost:8082/?wc-ajax=wcpay_create_payment_intent

**More expectations from the testing:** 

- Please test for both legacy UPE and split UPE (but no need for deferred intent UPE).
- Check the latest WC Log, and confirm seeing a message like this: 
```
2023-06-08T10:51:11+00:00 ERROR UPE Process Redirect Payment - Order ID mismatched. Received: 999. Intent Metadata Value: 1000
2023-06-08T10:51:11+00:00 INFO Error: We're not able to process this payment. Please try again later.
``` 
- In the example above, order 999 will not be marked as paid (through the sync process in the check-out page), while order 1000 will be marked as paid (through the webhook async processing)

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

*

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

htdat committed:
Add code to prevent the duplicate orders for the same payment intent.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6487` or branch name `fix/5786-avoid-multiple-order-upe-flow` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=fix/5786-avoid-multiple-order-upe-flow) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 20a124fc0ca14ccedfe2bdfde82fe077f1a61d99
- Build time: 2023-07-19 12:33:09 UTC

_Note: the build is updated when a new commit is pushed to this PR._

htdat committed:
Fix PHP tests.

htdat committed:
Avoid marking the order as failed when upe_process_redirect_order_id_mismatched is detected

htdat committed:
Add changelog

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow

htdat committed:
Add changelog and revert change in run-tests.sh

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow

RadoslavGeorgiev commented on includes/payment-methods/class-upe-payment-gateway.php:
```diff
@@ -795,13 +802,19 @@ public function process_redirect_payment( $order_id, $intent_id, $save_payment_m
+
+				/* translators: localized exception message */
+				$message = sprintf( __( 'UPE payment failed: %s', 'woocommerce-payments' ), $e->getMessage() );
+				$this->order_service->mark_payment_failed( $order, $intent_id, $status, $charge_id, $message );
```
Any reason to skip this? A meaningful note could provide context here, and AFAICT there are no downsides to calling `mark_payment_failed()`.

Was there a specific reason this was added?

htdat replied:
I've added this commit to explain why that's the case 824dad021cef1716b53172af3225ec4729c72e14

Basically, when we detect the order ID mismatch here, the provided intent ID is no longer relevant for the current processing order.

RadoslavGeorgiev reviewed (approved):
The code works as advertised! 🙇  :shipit: 

I have a single nitpick, could be ignored.

To confirm: The issue here is that the webhook comes in, and marks the order as completed, based on the outdated order ID within the payment intent? If yes, we might really need to rethink what we're doing with webhooks.

This doesn't prevent the PR from being shipped, but even testing this change is complicated. I don't imagine anyone going through this scenario during normal testing. We might want to open another issue, to address this differently (in addition to this check, which cannot do any harm).

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow

htdat committed:
Modify the Exception message.

- Differentiate it from a similar exception https://github.com/Automattic/woocommerce-payments/blob/a0b98f50ede90079d376d2521ff1aaa6043da454/includes/payment-methods/class-upe-payment-gateway.php#L745-L747

- Then, make it easy to trace back if this issue happens.

htdat committed:
Add a comment explaining why not marking order as failed when upe_process_redirect_order_id_mismatched exception is thrown.

htdat committed:
Add a flag to the redirect url when detecting the order ID mismatch.

htdat commented:
> To confirm: The issue here is that the webhook comes in, and marks the order as completed, based on the outdated order ID within the payment intent? If yes, we might really need to rethink what we're doing with webhooks.

@RadoslavGeorgiev - In the reversed way, actually, the order ID created in tab 2 per steps in this issue https://github.com/Automattic/woocommerce-payments/issues/5786 will be marked as paid. Webhook works correctly to do that. 

It's just that the sync process, i.e. method `UPE_Payment_Gateway::process_redirect_payment` is not working as expected, and this PR is to fix it. 

--- 

Also, I've added two other commits to improve this PR a bit too: a241b75a31a38e1e46fbc3d3e18d961e4cb91ca8 and 26ac73a9c07337f9260635128c13d228114ddbf5

It's ready to review again if necessary.

RadoslavGeorgiev reviewed (approved):
Great work, I understood things better. It's ready to 🚢  now :)

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow

naman03malhotra commented:
Blocked on #6575

naman03malhotra commented:
The target is to merge it by 6.2, 19th July

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow

# Conflicts:
#	includes/payment-methods/class-upe-payment-gateway.php

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow

htdat committed:
Apply the check to payment intents only.

htdat committed:
Add tests for validate_order_id_received_vs_intent_meta_order_id

htdat commented:
Ready for another review round @Automattic/gamma 

Changes include:

- ee48d85 - Based on the discussion in https://github.com/Automattic/woocommerce-payments/issues/6575#issuecomment-1614194510, I've modified to apply the new check method `validate_order_id_received_vs_intent_meta_order_id` for payment intents only. That's not ideal but it's acceptable as all the recognized issues so far are related to payment intents (see https://github.com/Automattic/woocommerce-payments/issues/5786). 
- 208411b - add new PHPUnit tests for the new method.

htdat commented:
This CI failure [E2E Tests - Pull Request / WC - latest | wcpay - merchant (pull_request)](https://github.com/Automattic/woocommerce-payments/actions/runs/5540262945/jobs/10112461112?pr=6487) **seems not** related to this PR. It's related to the dispute process instead:

> FAIL tests/e2e/specs/wcpay/merchant/merchant-disputes-submit-winning.spec.js (123.065 s)
>   Disputes > Submit winning dispute
>     ✕ should process and confirm a winning dispute (100010 ms)
> 
>   ● Disputes > Submit winning dispute › should process and confirm a winning dispute

All recent CI runs from other PRs in https://github.com/Automattic/woocommerce-payments/actions/workflows/e2e-pull-request.yml have failed with the same test suite too. 

_Update_: there is another PHP test failure [`WC_Payments_Test::test_ajax_init_woopay_sends_correct_customer_id`](https://github.com/Automattic/woocommerce-payments/actions/runs/5550007902/jobs/10134697555?pr=6487) but that's because of another PR - see p1689305403447819-slack-CGGCLBN58

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow

jessy-p commented:
Reviewed the latest code changes and LGTM!

htdat committed:
Merge branch 'develop' into fix/5786-avoid-multiple-order-upe-flow
